MdeModulePkg/Library/MemoryProtectionHobLib/DxeMemoryProtectionHobLib.c

```diff
@@ -171,6 +171,95 @@ inactive but RaiseErrorIfProtectionFails is active. RaiseErrorIfProtectionFails
       ));
     gDxeMps.ImageProtectionPolicy.Fields.RaiseErrorIfProtectionFails = 0;
   }
+
+  //
+  // the heap guard system does not support non-EFI_PAGE_SIZE alignments
+  // architectures that require larger RUNTIME_PAGE_ALLOCATION_GRANULARITY
+  // cannot have EfiRuntimeServicesCode, EfiRuntimeServicesData, EfiReservedMemoryType,
+  // and EfiACPIMemoryNVS guarded. OSes do not map guard pages anyway, so this is a
+  // minimal loss. Not guarding prevents alignment mismatches
+  //
+  if (RUNTIME_PAGE_ALLOCATION_GRANULARITY != EFI_PAGE_SIZE) {
+    if (gDxeMps.HeapGuardPageType.Fields.EfiACPIMemoryNVS != 0) {
+      DEBUG ((
+        DEBUG_WARN,
+        "%a: - RUNTIME_PAGE_ALLOCATION_GRANULARITY != EFI_PAGE_SIZE but Page Guard set on \
+        EfiACPIMemoryNVS. This is not supported by Heap Guard system, disabling.\n",
+        __func__
+        ));
+      gDxeMps.HeapGuardPageType.Fields.EfiACPIMemoryNVS = 0;
+    }
+
+    if (gDxeMps.HeapGuardPageType.Fields.EfiReservedMemoryType != 0) {
+      DEBUG ((
+        DEBUG_WARN,
+        "%a: - RUNTIME_PAGE_ALLOCATION_GRANULARITY != EFI_PAGE_SIZE but Page Guard set on \
+        EfiReservedMemoryType. This is not supported by Heap Guard system, disabling.\n",
+        __func__
+        ));
+      gDxeMps.HeapGuardPageType.Fields.EfiReservedMemoryType = 0;
+    }
+
+    if (gDxeMps.HeapGuardPageType.Fields.EfiRuntimeServicesCode != 0) {
+      DEBUG ((
+        DEBUG_WARN,
+        "%a: - RUNTIME_PAGE_ALLOCATION_GRANULARITY != EFI_PAGE_SIZE but Page Guard set on \
+        EfiRuntimeServicesCode. This is not supported by Heap Guard system, disabling.\n",
+        __func__
+        ));
+      gDxeMps.HeapGuardPageType.Fields.EfiRuntimeServicesCode = 0;
+    }
+
+    if (gDxeMps.HeapGuardPageType.Fields.EfiRuntimeServicesData != 0) {
+      DEBUG ((
+        DEBUG_WARN,
+        "%a: - RUNTIME_PAGE_ALLOCATION_GRANULARITY != EFI_PAGE_SIZE but Page Guard set on \
+        EfiRuntimeServicesData. This is not supported by Heap Guard system, disabling.\n",
+        __func__
+        ));
+      gDxeMps.HeapGuardPageType.Fields.EfiRuntimeServicesData = 0;
+    }
+
+    if (gDxeMps.HeapGuardPoolType.Fields.EfiACPIMemoryNVS != 0) {

```

Comment:
Can you also do a check against gDxeMps.HeapGuardPolicy.Fields.UefiPoolGuard (the global pool guard toggle) before printing the pool guard warnings?

Reply:
I can do this for the page and pool guards. I thought about doing it and had decided against it as it will already print if you don't have the toggle on but have any fields set and I wanted to make it clear on just one run that your settings were incorrect, i.e. not have it be that you run it, see oops I don't have the guard enabled, then rerun it and see oh, those memory types aren't valid for this platform. But I don't have any problem with checking it